### PR TITLE
Final Group Capstone Project

### PR DESCRIPTION
**Pull request by mavericks-db**
## Final Group Capstone Project

*** [FrontEnd Pull Request Link](https://github.com/jssol/bookracers-frontend/pull/24) ***
[Deployed Link - FrontEnd](https://effortless-starburst-07cb77.netlify.app/)

*** [BackEnd Pull Request Link](https://github.com/mavericks-db/BookRacers-Rails-Back-End/pull/34) ***
[Deployed Link -BackEnd](https://pacific-garden-78390.herokuapp.com/)

#### Basics
- You should follow the layout of [the provided design](https://www.behance.net/gallery/26425031/Vespa-Responsive-Redesign). You should change only the titles, descriptions, and photos - in order to create a website about something other than motorcycles.
- Select a theme for your website - is it going to be a website for booking doctor appointments, booking online classes, or something else?

#### Features

**Core features - for teams of each size**
- The user logs in to the website, only by typing the username (a proper authenticated login is a requirement if your group is made of 5 people).
- In the navigation panel, the user can see links to:
    - Motorcycles/doctors/classes/items that you selected as a theme.
    - "Reserve" form.
    - "My reservations".
    - "Add motorcycle/doctor/class/item that you selected as a theme" (the link is visible to everybody).
    - "Delete motorcycle/doctor/class/item that you selected as a theme" (the link is visible to everybody).
- On the main page, the user can see a [list of motorcycles/doctors/classes/items that you selected as a theme](https://www.behance.net/gallery/26425031/Vespa-Responsive-Redesign/modules/173005577).
    <p align="center">
      list</p>
- When the user selects a specific item, they can see the [details page](https://www.behance.net/gallery/26425031/Vespa-Responsive-Redesign/modules/173005579) with its full description (skip the "Rotate image" button).
    - In the details page, the user can click the "Reserve" button (in the design you can see the "Configure" button - please replace it with the "Reserve" button).
    <p align="center">
      details</p>

- When the user clicks the "Add item" link in the navigation panel they can see a form for adding a new item.

- Make the app responsive, creating both mobile and desktop versions.

**If your team has 3-4 members these are the required features; otherwise, consider them as optional features.**
- When the user clicks the "Delete item" link in the navigation panel they can see a list of all items with title and "Delete" button.
    - When the user clicks the "Delete" button, the selected item is marked as removed and does not show on the main list anymore.
- To reserve an appointment, the user has to select a date and city (username and selected item are autofilled).
    - Use the design based on the ["Book a vespa test-ride"](https://www.behance.net/gallery/26425031/Vespa-Responsive-Redesign/modules/173005583) and add all necessary inputs.
    - The user can also access the "Reserve" page from the navigation panel. In that case only username is autofilled.
    <p align="center">
     booking</p>

- When the user clicks the "My reservations" link in the navigation panel they can see a list of their reservations (with information about item name, date and city).
- Add full documentation for your API.

**If your team has 5 members these are the required features; otherwise, consider them as optional features.**

- Implement proper user authentication from the front-end to the server.
- Make sure that the "Add item" and "Delete item" links are accessible only by users who are admins.

#### Technical set up
- Set up the repository/repositories on GitHub and use Gitflow.
- You should use Postgres as your database
- Use Rails to create backend API.
- Use React & Redux to create frontend UI.
- You can choose if you want to set up your project as two separate apps or as one.
- The [Creative Commons license of the design](https://creativecommons.org/licenses/by-nc/4.0/) requires that you give appropriate credit to the author. Therefore, you must do it in the README of your project.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,24 +1,105 @@
...
+`bundle install`</br>
+`rails s`</br>
+Open a browser and go to: `https://localhost:3000` to view the app.
+
+## Testing
+
+`cd BookRacers`</br>
+run `rspec`
+
+## [Go Live]()
```
Comment (Meltrust) [inline README.md:47]: - [ ] This link is broken.  Please, consider updating it.

File: README.md [hunk 2]
```diff
@@ -1,24 +1,105 @@
...
 
-Things you may want to cover:
+- This is a FullStack Group project and here is the backend service provider for the BookRacer App which helps the frontend users to register(sign up) and login (sign in) using JWT. It also provides API endpoints to read, write, update, and delete data.
 
-* Ruby version
+- We also built a simple documentation for this project so users can have an idea on how to use the app.
 
-* System dependencies
+*** [Link to FrontEnd App](https://github.com/jssol/bookracers-frontend) ***
+*** [Link to Live Demo]() ***
```
Comment (Meltrust) [inline README.md:10]: - [ ] This link to the live demo is broken, kindly consider updating it.

File: README.md [hunk 3]
```diff
@@ -1,24 +1,105 @@
...
+- [x] And some other open source [ruby gems](https://rubygems.org/)
+
+## Usage
+
+Run the following commands:</br>
+`git clone https://github.com/jssol/bookracers-frontend.git`</br>
+`cd b\BookRacers`</br>
+`bundle install`</br>
+`rails s`</br>
+Open a browser and go to: `https://localhost:3000` to view the app.
```
Comment (Meltrust) [inline README.md:40]: - [ ] Please, kindly add the `rails db:create`, `rails db:migrate`, and the `rails db:seed` that your users will need to run and seed the project database.

File: spec/models/category_spec.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
+require 'rails_helper'
```
Comment (Meltrust) [inline spec/models/category_spec.rb:1]: - [ ] When running the test suite, I got 15 failing ones:

image

Kindly consider correcting these test, to ensure your app's integrity. 💪

File: test/channels/application_cable/connection_test.rb [hunk 5]
```diff
@@ -1,4 +1,4 @@
-require "test_helper"
+require 'test_helper'
```
Comment (Meltrust) [inline test/channels/application_cable/connection_test.rb:1]: - [ ] You are using Rspec, but the `test` folder is used by another test framework.  Please consider deleting the `test` folder if not being used.

File: Gemfile [hunk 6]
```diff
@@ -34,39 +22,48 @@ gem "jbuilder"
...
+# Use Rack CORS for handling Cross-Origin Resource Sharing (CORS), making cross-origin AJAX possible
+gem 'rack-cors'
+
+gem 'rubocop'
+
+gem 'jwt'
+
+gem 'active_model_serializers'
+
+gem 'cancancan'
```
Comment (Meltrust) [inline Gemfile:45]: - [ ] Some gems in your `gemfile` are not being used.  It seems to be the case with `cancancan` for instance.  Please, kindly remove the gems that are obsolete for the project.

That way, the dependency overhead will be reduced 💪

File: app/controllers/api/v1/reservations_controller.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+      render json: { error: 'Unable to find your reservation' }, status: :unprocessable_entity
+    end
+  end
+
+  def create
+    # @motorcycle = Motorcycle.find(params[:id])
+    @reservation = Reservation.new(reservation_params)
+    # @reservation.motorcycle_id = @motorcycle.id
+    # @reservation.user_id = current_user.id
+    # @reservation.total_price =
```
Comment (Meltrust) [inline app/controllers/api/v1/reservations_controller.rb:21]: - [ ] Please, kindly remove these old pieces of old commented code, to improve the cleanliness of the repo.  💪

**Review by Meltrust — CHANGES_REQUESTED** [review]
Hi @mavericks-db ,

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation, but you are almost there!






### To highlight:

-  Nice seeds✔️
-  App is working ✔️ 
-  Nice routing ✔️
-  Nice use of active storage ✔️ 

You can do it

## Suggested Changes ♻️
_Check the comments under the review._

You can use as many of my suggestions as you want.  If there is anything you would like to skip - feel free to do that.  However, I strongly recommend you take them into account as they can make your code better.


Cheers and Happy coding!👏👏👏

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------

File: README.md (hunk 1 above)
Comment (mavericks-db) [reply README.md:47]: Done. Thanks.

File: README.md (hunk 2 above)
Comment (mavericks-db) [reply README.md:10]: Done. Thanks.

File: README.md (hunk 3 above)
Comment (mavericks-db) [reply README.md:40]: Done. Thanks.

File: Gemfile (hunk 6 above)
Comment (mavericks-db) [reply Gemfile:45]: Done. Thanks.

File: app/controllers/api/v1/reservations_controller.rb (hunk 7 above)
Comment (mavericks-db) [reply app/controllers/api/v1/reservations_controller.rb:21]: Removed. Thanks.

File: README.md (hunk 3 above)
Comment (mavericks-db) [reply README.md:40]: Added. Thanks.

File: test/channels/application_cable/connection_test.rb (hunk 5 above)
Comment (mavericks-db) [reply test/channels/application_cable/connection_test.rb:1]: Deleted. Thanks.

**Review by mcrd25 — APPROVED** [review]
Hi Team, 

In my opinion, your project is ready for the presentation!  There is nothing else to say other than... it's time to merge it :shipit: 
*A tip as you move unto to real world projects*
I'd recommend to make your tests better in the future, maybe you can get more specific with your tests - what do you mean for something to be valid? Is the message you are getting what you expect? What about when it's not valid, id the message you are getting what you expect? - it can be not valid and still not behave in the way you expect it to.


Good luck! 🎉 


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**Comment by mwenyoa** [comment]
Thanks @mavericks-db for having worked on the required changes and Thanks @mcrd25 for taking time for reviewing our project and giving us resource feedback that we can use as we build real life projects.
 Cheers and Happy coding.

**Pull request merged**